akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowRecoverWithSpec.scala

```diff
@@ -0,0 +1,103 @@
+
+  "A RecoverWith" must {
+    "recover when there is a handler" in assertAllStagesStopped {
+      Source(1 to 4).map { a ⇒ if (a == 3) throw ex else a }
+        .recoverWith { case t: Throwable ⇒ Source(List(0, -1)) }
+        .runWith(TestSink.probe[Int])
+        .request(2)
+        .expectNextN(1 to 2)
+        .request(1)
+        .expectNext(0)
+        .request(1)
+        .expectNext(-1)
+        .expectComplete()
+    }
+
+    "cancel susbream if parent is trminated when there is a handler" in assertAllStagesStopped {
+      Source(1 to 4).map { a ⇒ if (a == 3) throw ex else a }
+        .recoverWith { case t: Throwable ⇒ Source(List(0, -1)) }
+        .runWith(TestSink.probe[Int])
+        .request(2)
+        .expectNextN(1 to 2)
+        .request(1)
+        .expectNext(0)
+        .cancel()
+    }
+
+    "failed stream if handler is not for such exception type" in assertAllStagesStopped {
+      Source(1 to 3).map { a ⇒ if (a == 2) throw ex else a }
+        .recoverWith { case t: IndexOutOfBoundsException ⇒ Source.single(0) }
+        .runWith(TestSink.probe[Int])
+        .request(1)
+        .expectNext(1)
+        .request(1)
+        .expectError(ex)
+    }
+
+    "not influence stream when there is no exceptions" in assertAllStagesStopped {
+      Source(1 to 3).map(identity)
+        .recoverWith { case t: Throwable ⇒ Source.single(0) }
+        .runWith(TestSink.probe[Int])
+        .request(3)
+        .expectNextN(1 to 3)
+        .expectComplete()
+    }
+
+    "finish stream if it's empty" in assertAllStagesStopped {
+      Source.empty.map(identity)
+        .recoverWith { case t: Throwable ⇒ Source.single(0) }
+        .runWith(TestSink.probe[Int])
+        .request(3)
+        .expectComplete()
+    }
+
+    "switch the second time if alternative source throws exception" in assertAllStagesStopped {
+      val k = Source(1 to 3).map { a ⇒ if (a == 3) throw new IndexOutOfBoundsException() else a }
+        .recoverWith {
+          case t: IndexOutOfBoundsException ⇒
+            Source(List(11, 22)).map(m ⇒ if (m == 22) throw new IllegalArgumentException() else m)
+          case t: IllegalArgumentException ⇒ Source(List(33, 44))
+        }.runWith(TestSink.probe[Int])
+        .request(2)
+        .expectNextN(List(1, 2))
+        .request(2)
+        .expectNextN(List(11, 33))
+        .request(1)
+        .expectNext(44)
+        .expectComplete()
+    }
+
+    "terminate with exception if altrnative source failed" in assertAllStagesStopped {
+      Source(1 to 3).map { a ⇒ if (a == 3) throw new IndexOutOfBoundsException() else a }
+        .recoverWith {
+          case t: IndexOutOfBoundsException ⇒
+            Source(List(11, 22)).map(m ⇒ if (m == 22) throw ex else m)
+        }.runWith(TestSink.probe[Int])
+        .request(2)
+        .expectNextN(List(1, 2))
+        .request(1)
+        .expectNextN(List(11))
+        .request(1)
+        .expectError(ex)
+    }
+  }
+}
```

Comment:
Ok, so if I have the following setup:

```scala
val src = Source.single(1) ++ Source.failed(ex) // this does not work because concat is eager, but you get what I mean

src.recoverWith(_ => src) // This will be equivalent to Source.repeated(1)?
```